nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/QueryCassandra.java

```diff
@@ -191,76 +203,110 @@ public void onScheduled(final ProcessContext context) {
+                                            out, charset, queryTimeout, TimeUnit.MILLISECONDS));
+                                }
+                            } else {
+                                if (AVRO_FORMAT.equals(outputFormat)) {
+                                    nrOfRows.set(convertToAvroStream(resultSet, maxRowsPerFlowFile,
+                                            out, 0, null));
+                                } else if (JSON_FORMAT.equals(outputFormat)) {
+                                    nrOfRows.set(convertToJsonStream(resultSet, maxRowsPerFlowFile,
+                                            out, charset, 0, null));
+                                }
                             }
-                        }
 
-                    } catch (final TimeoutException | InterruptedException | ExecutionException e) {
-                        throw new ProcessException(e);
+                        } catch (final TimeoutException | InterruptedException | ExecutionException e) {
+                            throw new ProcessException(e);
+                        }
                     }
-                }
-            });
+                });
+
 
-            // set attribute how many rows were selected
-            fileToProcess = session.putAttribute(fileToProcess, RESULT_ROW_COUNT, String.valueOf(nrOfRows.get()));
+                // set attribute how many rows were selected
+                fileToProcess = session.putAttribute(fileToProcess, RESULT_ROW_COUNT, String.valueOf(nrOfRows.get()));
 
-            // set mime.type based on output format
-            fileToProcess = session.putAttribute(fileToProcess, CoreAttributes.MIME_TYPE.key(),
-                    JSON_FORMAT.equals(outputFormat) ? "application/json" : "application/avro-binary");
+                // set mime.type based on output format
+                fileToProcess = session.putAttribute(fileToProcess, CoreAttributes.MIME_TYPE.key(),
+                        JSON_FORMAT.equals(outputFormat) ? "application/json" : "application/avro-binary");
 
-            logger.info("{} contains {} Avro records; transferring to 'success'",
-                    new Object[]{fileToProcess, nrOfRows.get()});
-            session.getProvenanceReporter().modifyContent(fileToProcess, "Retrieved " + nrOfRows.get() + " rows",
-                    stopWatch.getElapsed(TimeUnit.MILLISECONDS));
-            session.transfer(fileToProcess, REL_SUCCESS);
+                logger.info("{} contains {} records; transferring to 'success'",
+                        new Object[]{fileToProcess, nrOfRows.get()});
+                session.getProvenanceReporter().modifyContent(fileToProcess, "Retrieved " + nrOfRows.get() + " rows",
+                        stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+                session.transfer(fileToProcess, REL_SUCCESS);
+                session.commit();
```

Comment:
This is kind of a mashup of two features, Max Rows Per FlowFile and Output Batch Size (at least in terms of the SQL version of these features). The former is to limit the number of rows in each flow file, often used (without Output Batch Size) if some processor downstream needs to read the entire content into memory. The latter (Output Batch Size) is the number of flow files to commit, and is used to "get things moving downstream" by actually sending the flow files to the outgoing connection(s), rather than just marking them for transfer and actually sending them only when the entire result set has been processed (no matter how many rows per flow file).

My preference is to keep these features separate, but I'm happy to discuss if you don't see the need for having the choice between either, both, or none. I just think it makes it more flexible for the user.

Reply:
The "Issue" here is as we have this two features, the "Retrieved" attribute has a new semantic: for "Output Batch Size" is > 0, "Retrieved" means the number of rows on this flow file, if "Output Batch Size" is == 0, "Retrieved" means the total number of rows returned by the query, independently the number of output files.

My particular use case demands these parameters be segregated, because our database has millions of rows, and wait for the query end will delay the further processing, and the "Retrieved" attribute lost his original semantics, but in the case where the user set "Output Batch Size" = 0, the original semantic remain unchanged.

On this PR these features are segregated, but the "Retrieved" attribute is a little confusing.

Reply:
Hi @mattyb149 @! Did you have a chance to look at my comments? What do you think?

Reply:
By the "retrieved attribute", do you mean `RESULT_ROW_COUNT`? I wasn't addressing that per se; rather I am noting that the session seems to always be committed after processing Max Rows Per Flow File. That's what I meant by the "mashup" of the two properties that exist in other processors. It is possible in those processors to specify Max Rows Per Flow File, but still have all the flow files transferred at once (i.e. after all the rows have been processed). That allows you to have smaller flow files for downstream processing, but you can still add attributes such as total count, fragment index, etc.

I realize that doesn't solve your use case, where you want the flow files to be sent downstream while the rows are still being processed. That's why there's a separate property Output Batch Size in those processors, and if set to X will commit the session after X number of flow files have been transferred.

In this case you have one property and you're kind of using that value for both the values of the analogous properties in the other processors. I'm just suggesting that you have two separate properties, for consistency and flexibility. For your use case I believe you'll just always set Output Batch Size to 1.

Reply:
I think that I finally understand your point @mattyb149, thanks for the clarifications!